gmail/src/models/project.ts

```diff
@@ -0,0 +1,69 @@
+import { postJsonRpc } from "../utils/http";
+import { URLS } from "../const";
+import { ErrorMessage } from "../models/error_message";
+import { State } from "../models/state";
+
+/**
+ * Represent a "project.project" record.
+ */
+export class Project {
+    id: number;
+    name: string;
+
+    /**
+     * Unserialize the project object (reverse JSON.stringify).
+     */
+    static fromJson(values: any): Project {
+        const project = new Project();
+        project.id = values.id;
+        project.name = values.name;
+        return project;
+    }
+
+    /**
+     * Parse the dictionary return by the Odoo endpoint.
+     */
+    static fromOdooResponse(values: any): Project {
+        const project = new Project();
+        project.id = values.project_id;
+        project.name = values.name;
+        return project;
+    }
+
+    /**
+     * Make a RPC call to the Odoo database to search a project.
+     */
+    static searchProject(query: string): [Project[], ErrorMessage] {
+        const url = State.odooServerUrl + URLS.SEARCH_PROJECT;
+        const accessToken = State.accessToken;
+
+        const response = postJsonRpc(url, { search_term: query }, { Authorization: "Bearer " + accessToken });
+
+        if (!response) {
+            return [[], new ErrorMessage("http_error_odoo")];
+        }
+
+        return [response.map((values: any) => Project.fromOdooResponse(values)), new ErrorMessage()];
+    }
+
+    /**
+     * Make a RPC call to the Odoo database to create a project
+     * and return the newly created record.
+     */
+    static createProject(name: string): Project {
+        const url = State.odooServerUrl + URLS.CREATE_PROJECT;
+        const accessToken = State.accessToken;
+
+        const response = postJsonRpc(url, { name: name }, { Authorization: "Bearer " + accessToken });
+
+        const projectId = response ? response.project_id || null : null;
```

Comment:
Technically route returns project_id and name, in case someone overrided create behavior to add some automatic stuff to name could be great to take this value instead of the given name. That way we are consistent with what is stored in DB.

Reply:
Good idea :)